measurement/plugins/ip_route/measurements.py

```diff
@@ -0,0 +1,119 @@
+
+"""
+import socket
+import validators
+from validators import ValidationFailure
+
+# Both of these imports seem to be required in order to mock the traceroute function
+import scapy
+from scapy.layers.inet import traceroute
+
+
+from measurement.measurements import BaseMeasurement
+from measurement.results import Error
+from measurement.plugins.ip_route.results import IPRouteMeasurementResult
+from measurement.plugins.latency.measurements import LatencyMeasurement
+
+ROUTE_ERRORS = {
+    "route-err": "iproute encountered an unknown error",
+    "route-address": "iproute failed to find socket address",
+    "route-permission": "iproute does not have permission to create a raw socket",
+}
+
+
+class IPRouteMeasurement(BaseMeasurement):
+    def __init__(self, id, hosts, route_timeout=10, count=4):
+        super(IPRouteMeasurement, self).__init__(id=id)
+
+        if len(hosts) < 1:
+            raise ValueError("At least one host must be provided.")
+        for host in hosts:
+            validated_domain = validators.domain(host)
+            validated_ip = validators.ipv4(host)
+            if isinstance(validated_domain, ValidationFailure) & isinstance(
+                validated_ip, ValidationFailure
+            ):
+                raise ValueError("`{host}` is not a valid host".format(host=host))
+
+        if count < 0:
+            raise ValueError(
+                "A value of {count} was provided for the number of pings. This must be a positive "
+                "integer or `0` to turn off the ping.".format(count=count)
+            )
+
+        self.id = id
+        self.hosts = hosts
+        self.route_timeout = route_timeout
+        self.count = count
+
+    def measure(self):
+        initial_latency_results = self._find_least_latent_host(self.hosts)
+        least_latent_host = initial_latency_results[0][0]
+        results = [self._get_traceroute_result(least_latent_host)]
+        if self.count > 0:
+            latency_measurement = LatencyMeasurement(
+                self.id, least_latent_host, count=self.count
+            )
+            results.append(latency_measurement.measure()[0])
+        results.extend([res for _, res in initial_latency_results])
+        return results
+
+    def _find_least_latent_host(self, hosts):
+        """
+        Performs a latency test for each specified host
+        Returns a sorted list of LatencyResults, sorted by average latency
+        """
+        initial_latency_results = []
+        for host in hosts:
+            latency_measurement = LatencyMeasurement(self.id, host, count=2)
+            initial_latency_results.append((host, latency_measurement.measure()[0]))
+        return sorted(
+            initial_latency_results,
+            key=lambda x: (x[1].average_latency is None, x[1].average_latency),
+        )
+
+    def _get_traceroute_result(self, host):
+        try:
+            traceroute_out = scapy.layers.inet.traceroute(host)
```

Comment:
When running without `sudo` there's an uncaught exception which is [apparently expected](https://github.com/secdev/scapy/issues/1942). We should catch it and return an error in the standard way using `results.Error`.

```
In [1]: from measurement.plugins.ip_route.measurements import IPRouteMeasurement                                                                                                                                   

In [2]: import uuid                                                                                                                                                                                                

In [3]: irm = IPRouteMeasurement( 
   ...:     "{uuid}".format(uuid=uuid.uuid4()), 
   ...:     ["www.google.com", "www.facebook.com"] 
   ...:     )                                                                                                                                                                                                      

In [4]: r = irm.measure()                                                                                                                                                                                          
Exception ignored in: <function SuperSocket.__del__ at 0x7fb5f00f0bf8>
Traceback (most recent call last):
  File "/home/james/.pyenv/versions/3.7.3/envs/honestybox-measurement/lib/python3.7/site-packages/scapy/supersocket.py", line 123, in __del__
    self.close()
  File "/home/james/.pyenv/versions/3.7.3/envs/honestybox-measurement/lib/python3.7/site-packages/scapy/arch/linux.py", line 481, in close
    set_promisc(self.ins, self.iface, 0)
AttributeError: 'L3PacketSocket' object has no attribute 'ins'
```

Reply:
Huh, thats weird, only ever got a `PermissionError` on my machine, thanks for the heads up!

Reply:
Actually I just noticed this is the first line:

`Exception ignored in: <function SuperSocket.__del__ at 0x7f5f44078268>`

Looks like the ignoring of the exception is normal behaviour in `__del__` as per the [docs](https://docs.python.org/2/reference/datamodel.html#object.__del__). Note that _..."a warning is printed to sys.stderr instead"_ which I guess is what we're seeing here.

I'd like to understand why this is happening and whether or not it can be avoided. Really we should be in complete control of errors and how they're handled, including anything that's spat out to stderr or stdout.